clients/src/test/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporterTest.java

```diff
@@ -0,0 +1,573 @@
+            .telemetrySender()).state());
+    }
+
+    @Test
+    public void testTelemetryReporterCloseMultipleTimesNoException() {
+        clientTelemetryReporter.close();
+        clientTelemetryReporter.close();
+        assertEquals(ClientTelemetryState.TERMINATED, ((ClientTelemetryReporter.DefaultClientTelemetrySender) clientTelemetryReporter
+            .telemetrySender()).state());
+    }
+
+    @Test
+    public void testTelemetrySenderTimeToNextUpdate() {
+        ClientTelemetryReporter.DefaultClientTelemetrySender telemetrySender = (ClientTelemetryReporter.DefaultClientTelemetrySender) clientTelemetryReporter.telemetrySender();
+
+        assertEquals(ClientTelemetryState.SUBSCRIPTION_NEEDED, telemetrySender.state());
+        assertEquals(0, telemetrySender.timeToNextUpdate(100));
+
+        telemetrySender.updateSubscriptionResult(subscription, time.milliseconds());
+        assertEquals(20000, telemetrySender.timeToNextUpdate(100), 200);
+
+        assertTrue(telemetrySender.maybeSetState(ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS));
+        assertEquals(100, telemetrySender.timeToNextUpdate(100));
+
+        assertTrue(telemetrySender.maybeSetState(ClientTelemetryState.PUSH_NEEDED));
+        long time = telemetrySender.timeToNextUpdate(100);
+        assertTrue(time > 0 && time >= 0.5 * time && time <= 1.5 * time);
+
+        assertTrue(telemetrySender.maybeSetState(ClientTelemetryState.PUSH_IN_PROGRESS));
+        assertEquals(100, telemetrySender.timeToNextUpdate(100));
+
+        assertTrue(telemetrySender.maybeSetState(ClientTelemetryState.TERMINATING_PUSH_NEEDED));
+        assertEquals(0, telemetrySender.timeToNextUpdate(100));
+
+        assertTrue(telemetrySender.maybeSetState(ClientTelemetryState.TERMINATING_PUSH_IN_PROGRESS));
+        assertEquals(Long.MAX_VALUE, telemetrySender.timeToNextUpdate(100));
+
+        assertTrue(telemetrySender.maybeSetState(ClientTelemetryState.TERMINATED));
+        assertThrows(IllegalStateException.class, () -> telemetrySender.timeToNextUpdate(100));
+    }
+
+    @Test
+    public void testCreateRequestSubscriptionNeeded() {
+        ClientTelemetryReporter.DefaultClientTelemetrySender telemetrySender = (ClientTelemetryReporter.DefaultClientTelemetrySender) clientTelemetryReporter.telemetrySender();
+        assertEquals(ClientTelemetryState.SUBSCRIPTION_NEEDED, telemetrySender.state());
+
+        Optional<AbstractRequest.Builder<?>> requestOptional = telemetrySender.createRequest();
+        assertNotNull(requestOptional);
+        assertTrue(requestOptional.isPresent());
+        assertTrue(requestOptional.get().build() instanceof GetTelemetrySubscriptionsRequest);
+        GetTelemetrySubscriptionsRequest request = (GetTelemetrySubscriptionsRequest) requestOptional.get().build();
+
+        GetTelemetrySubscriptionsRequest expectedResult = new GetTelemetrySubscriptionsRequest.Builder(
```

Comment:
Hey I just realized - we should file a separate jira to test different request versions.  This helps to verify the correctness of the request. It's really simple, basically you just need to verify the fields are correct.  There's an example test `public void testValidateConsumerGroupHeartbeatRequest(final short version)`

Reply:
Filed: https://issues.apache.org/jira/browse/KAFKA-15958